src/core/labeling/qgspallabeling.cpp

```diff
@@ -2033,6 +2033,7 @@ void QgsPalLayerSettings::registerFeature( const QgsFeature &f, QgsRenderContext
       {
         quadOff = static_cast< QuadrantPosition >( quadInt );
         ddFixedQuad = true;
+        anchorPosition = geom.centroid().asPoint();
```

Comment:
Shouldn't that be moved out of the isValid() branch? If the expression is not valid for the current feature won't it fallback to a default and still be offsetQuad based?

Reply:
Possibly? I've been looking at where the anchorPosition is actually utilised (which is only in the dxf export code), and the logic is tricky to understand. It's using the anchor position IF a data defined offset quad is set, but without first checking that the data defined value is valid. I don't really understand this, and I'm struggling to understand why we would be using the anchor position in this case at all. Shouldn't the anchor position ONLY be used for fixed position labels OR for ALL labels set to a quadrant mode, regardless of whether or not these use data defined quadrant offsets?

Reply:
I'll come up with an improved version of this code.
I found the pal code hard to understand on my end, that's probably the reason I failed to replicate correctly.

Basically, if offset quad (dd or not) or hali/vali is used (which is always dd if I got that right, and should always have a dd fixed position, even though the enabled state on the config page that reflects that could need some :heart: ), the anchor point needs to be present.

Reply:
https://github.com/qgis/QGIS/pull/33974